addons/portal/controllers/mail.py

```diff
@@ -144,6 +160,102 @@ def portal_message_fetch(self, res_model, res_id, domain=False, limit=10, offset
             'message_count': Message.search_count(domain)
         }
 
+    @http.route('/portal/binary/upload_attachment', type='http', auth="public")
+    def upload_attachment(self, callback, model, id, ufile, **kwargs):
+        files = request.httprequest.files.getlist('ufile')
+        out = """<script language="javascript" type="text/javascript">
+                    var win = window.top.window;
+                    win.jQuery(win).trigger(%s, %s);
+                </script>"""
+
+        access_token = kwargs.get('access_token')
+        res_id = int(id)
+        args = []
+
+        try:
+            document = request.env[model].browse(res_id)
+            document_sudo = document.sudo().exists()
+            if not document_sudo:
+                raise MissingError("This document does not exist.")
+            try:
+                document.check_access_rights('read')
+                document.check_access_rule('read')
+            except AccessError:
+                if not access_token or not _check_special_access(model, res_id, token=access_token):
+                    raise Forbidden()
+        except (AccessError, MissingError):
+            raise Forbidden()
+
+        if len(files) > MAX_FILE_TO_UPOLAD:
+            args.append({'error': _("Oops! You can not upload more than 10 files.")})
+            return out % (json.dumps(callback), json.dumps(args))
+
+        AttachmentSudo = request.env['ir.attachment'].sudo()
+
+        for ufile in files:
+            data = ufile.read()
+            file_size = len(data) * 3 / 4
+            if (file_size / 1024.0 / 1024.0) > MAX_FILE_TO_UPOLAD:
+                args.append({'error': _("Oops! You can not upload a file larger than 25MB.")})
+
+            filename = ufile.filename
+            if request.httprequest.user_agent.browser == 'safari':
+                # Safari sends NFD UTF-8 (where é is composed by 'e' and [accent])
+                # we need to send it the same stuff, otherwise it'll fail
+                filename = unicodedata.normalize('NFD', ufile.filename)
+            try:
+                attachment = AttachmentSudo.create({
+                    'name': filename,
+                    'datas': base64.encodebytes(data),
+                    'datas_fname': filename,
+                    'res_model': model,
```

Comment:
When doing a two step upload (upload attachments then giving ids in message post) the attachment should be added on mail.compose.message with id 0. Message post will add it to the record automatically. This will avoid to add attachment to record if user cancel message writing.